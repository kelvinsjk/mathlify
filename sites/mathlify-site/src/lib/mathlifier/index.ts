/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
