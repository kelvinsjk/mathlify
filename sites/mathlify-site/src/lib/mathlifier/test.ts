import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
