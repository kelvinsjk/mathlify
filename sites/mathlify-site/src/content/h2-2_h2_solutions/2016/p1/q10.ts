import { Answer } from '$lib/classes/answer';
import { fnTerm, sum } from 'mathlify';

import { sqrtInverse } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { generateFn } from '$content/h2-1_h2_learn/01_functions/01_concepts/02_domain-and-range/02_practice-1';
import { EquationWorking } from 'mathlify/working';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { simplifySurd } from 'mathlify/fns';
import { bisection, cubicRoot } from 'mathlify/numerical';
import { QED } from '$lib/utils/typesetting/utils';

export const answer = new Answer();

// 1 + sqrt(x)

const [a, b] = [0, 1];

const state: Parameters<typeof sqrtInverse>[0] = {
	fnType: 'sqrt',
	a,
	b,
	restriction: false,
	// dummies
	definition: false,
	c: 1,
	unknownConstants: false
};
const f = generateFn(state)[1];

// ai
{
	const { ans, soln } = sqrtInverse(state, f, { definition: false });
	answer.addSubPart(ans, soln);
}
// aii
{
	const working = new EquationWorking(fnTerm('f', f), 'x');
	EquationWorking.RegisterCustomSimplifier(simplifySurd);
	working.subIn({ f }).moveTerms(1).square().expand().moveTerms(1);
	working.square().expand().moveTerms(0, { swapSides: true }).divide('x');
	EquationWorking.DeregisterCustomSimplifier();
	const cubic = working.eqn.lhs;

	const soln1 = mathlifier`$${'gather*'} ff(x) = x \\\\ ${working}`;
	const x2 = cubicRoot(cubic);
	const x1 = bisection(cubic.fn, 0, 1);
	const x3 = bisection(cubic.fn, 2, 3);

	const soln2 = mathlifier`Solving with a GC,

$${{}} x = ${x1.toPrecision(3)}, \\quad ${x2} \\quad \\text{or} \\quad ${x3.toPrecision(3)}

From ${{}} {\\sqrt{\\sqrt{x}+1}+1=x,}
we observe that ${{}} {x>1.}
Hence

$${{}} x = ${x3.toPrecision(3)} \\; ${QED} \n\n`;
	const soln3 = mathlifier`From ${{}}{ff(x)=x},
we can apply ${'f^{-1}'}
on both sides

$${'align*'} ff(x) &= x
\\\\ f^{-1}ff(x) &= f^{-1}(x)
\\\\ f(x) &= f^{-1}(x)

Hence the value of ${'x'}
obtained satisfies the equation ${{}} {f(x)=f^{-1}(x)} \\; ${QED}`;
	const ans = mathlifier`${{}} x = ${x3.toPrecision(3)}.`;
	answer.addSubPart(ans, soln1 + soln2 + soln3);
}
answer.newPart();
// b
{
	const evenCase = sum(2, fnTerm('g', [[1, '/', 2], 'n']));
	const oddCase = sum(1, fnTerm('g', sum('n', -1)));
	const soln = mathlifier`$${'align*'} g(2) &= ${evenCase.subIn({ n: 2 })}
\\\\ &= 2 + ${oddCase.subIn({ n: 1 })}
\\\\ &= 3 + 1
\\\\ &= 4

$${'align*'} g(4) &= ${evenCase.subIn({ n: 4 })}
\\\\ &= 2 + 4
\\\\ &= 6 \\; ${QED}

$${'align*'} g(6) &= ${evenCase.subIn({ n: 6 })}
\\\\ &= 2 + ${oddCase.subIn({ n: 3 })}
\\\\ &= 3 + 4
\\\\ &= 7

$${'align*'} g(7) &= ${oddCase.subIn({ n: 7 })}
\\\\ &= 1 + 7
\\\\ &= 8 \\; ${QED}

$${'align*'} g(12) &= ${evenCase.subIn({ n: 12 })}
\\\\ &= 2 + ${evenCase.subIn({ n: 6 })}
\\\\ &= 2 + 7
\\\\ &= 9 \\; ${QED}
`;
	const ans = mathlifier`${{}} g(4) = 6.
\\
${{}} g(7) = 8.
\\
${{}} g(12) = 9.`;
	answer.addSubPart(ans, soln);

	// bii
	const soln2 = mathlifier`$${'align*'} g(8) &= ${evenCase.subIn({ n: 8 })}
\\\\ &= 2 + 6
\\\\ &= 8

Since ${{}} {g(7)=g(8)=8,}
${'g'}
is not a one-to-one function and ${'g'}
does not have an inverse ${QED}`;
	const ans2 = mathlifier`${'g'}
does not have an inverse as ${{}}{g(7)=g(8)=8}
so ${'g'}
is not a one-to-one function.`;
	answer.addSubPart(ans2, soln2);
}
