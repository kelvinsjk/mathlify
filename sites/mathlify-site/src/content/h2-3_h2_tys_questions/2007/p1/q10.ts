/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Polynomial } from 'mathlify';

export const question = new Question();

const poly = new Polynomial([3, -5, 2], { variable: 'r' });

question.addBody(mathlifier`A geometric series has a common ratio ${'r'},
and an arithmetic series has first term ${'a'}
and common difference ${'d'},
where ${'a'}
and ${'d'}
are non-zero.

The first three terms of the geometric series are
equal to the first, fourth and sixth terms respectively of the arithmetic series`);

// a
question.addPart(mathlifier`Show that ${{}} {${poly}.}`, { marks: 4 });
// b
question.addPart(
	mathlifier`Deduce that the geometric series is convergent
and find, in terms of ${'a'},
the sum to infinity.`,
	{ marks: 5 }
);
// c
question.addPart(
	mathlifier`The sum of the first ${'n'}
terms of the arithmetic series is denoted by ${'S'}.
Given that ${{}} {a>0,}
find the set of possible values of ${'n'}
for which ${'S'}
exceeds ${'4a'}.`,
	{ marks: 5 }
);
