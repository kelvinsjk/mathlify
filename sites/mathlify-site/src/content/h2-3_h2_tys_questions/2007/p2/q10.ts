/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { quotient } from 'mathlify';

export const question = new Question();

const p = quotient(1, 8);

question.addBody(mathlifier`A player throws three darts at a target.
The probability that he is successful in hitting the target with
his first throw is ${p}.
For each of his second and third throws, the probability of success is

- twice the probability of success on the preceding throw if that throw was successful,
- the same as the probability of success on the preceding throw if that throw was unsuccessful.`);
// a
question.addPart(mathlifier`Construct a probability tree showing this information.`, { marks: 3 });
// b
question.addPart(mathlifier`Find`);
question.addSubPart(`the probability that all three throws are successful,`, { marks: 2 });
question.addSubPart(`the probability that at least two throws are successful,`, { marks: 2 });
question.addSubPart(
	`the probability that the third throw is successful
given that exactly two of the three throws are successful.`,
	{ marks: 4 }
);
