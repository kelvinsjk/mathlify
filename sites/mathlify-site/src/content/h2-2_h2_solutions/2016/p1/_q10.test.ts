import { expect, it } from 'vitest';
import { answer } from './q10';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			parts: [
				{
					body: mathlifier`${{}} f^{-1}(x) = \\left( x - 1 \\right)^2.
\\
${{}} D_{f^{-1}} = R_{f} = \\left[ 1, \\infty \\right).`
				},
				{ body: mathlifier`${{}} x = 2.62.` }
			]
		},
		{
			parts: [
				{
					body: mathlifier`${{}} g(4) = 6.
\\
${{}} g(7) = 8.
\\
${{}} g(12) = 9.`
				},
				{
					body: mathlifier`${{}}g
does not have an inverse as ${{}}{g(7)=g(8)=8}
so ${'g'}
is not a one-to-one function.`
				}
			]
		}
	]
};

it('2016p1q10', () => {
	expect(answer).toMatchSnapshot();
	expect(ans).toEqual(actual);
});
