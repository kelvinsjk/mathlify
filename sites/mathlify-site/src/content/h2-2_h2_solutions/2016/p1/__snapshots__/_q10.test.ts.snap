// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2016p1q10 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "parts": [
          {
            "body": "$\`  f^{-1}(x) = \\left( x - 1 \\right)^2. \`
\\
$\`  D_{f^{-1}} = R_{f} = \\left[ 1, \\infty \\right). \`",
          },
          {
            "body": "$\`  x = 2.62. \`",
          },
        ],
      },
      {
        "parts": [
          {
            "body": "$\`  g(4) = 6. \`
\\
$\`  g(7) = 8. \`
\\
$\`  g(12) = 9. \`",
          },
          {
            "body": "$\` g \`
does not have an inverse as $\` {g(7)=g(8)=8} \`
so $\` g \`
is not a one-to-one function.",
          },
        ],
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "parts": [
          {
            "body": "$$\` \\begin{gather*} \\text{Let } y = \\sqrt{x} + 1
	\\\\ \\sqrt{x} = y - 1
	\\\\ x = \\left( y - 1 \\right)^2
\\end{gather*} \`$$\`  f^{-1}(x) = \\left( x - 1 \\right)^2 \\; \\blacksquare  \`

$$\` \\begin{align*} D_{f^{-1}} &= R_{f} \\\\ &= \\left[ 1, \\infty \\right) \\; \\blacksquare
\\end{align*} \`",
          },
          {
            "body": "$$\` \\begin{gather*} ff(x) = x \\\\ f\\left(\\sqrt{x} + 1\\right) = x
	\\\\ \\sqrt{\\sqrt{x} + 1} + 1 = x
	\\\\ \\sqrt{\\sqrt{x} + 1} = x - 1
	\\\\ \\sqrt{x} + 1 = \\left( x - 1 \\right)^2
	\\\\ \\sqrt{x} + 1 = x^2 - 2x + 1
	\\\\ \\sqrt{x} = x^2 - 2x
	\\\\ x = \\left( x^2 - 2x \\right)^2
	\\\\ x = x^4 - 4x^3 + 4x^2
	\\\\ x^4 - 4x^3 + 4x^2 - x = 0
	\\\\ x^3 - 4x^2 + 4x - 1 = 0
\\end{gather*} \`Solving with a GC,

$$\`  x = 0.382, \\quad 1 \\quad \\text{or} \\quad 2.62 \`

From $\`  {\\sqrt{\\sqrt{x}+1}+1=x,} \`
we observe that $\`  {x>1.} \`
Hence

$$\`  x = 2.62 \\; \\QED   \`

From $\` {ff(x)=x}, \`
we can apply $\` f^{-1} \`
on both sides

$$\` \\begin{align*} ff(x) &= x
\\\\ f^{-1}ff(x) &= f^{-1}(x)
\\\\ f(x) &= f^{-1}(x)
\\end{align*} \`

Hence the value of $\` x \`
obtained satisfies the equation $\`  {f(x)=f^{-1}(x)} \\; \\QED  \`",
          },
        ],
      },
      {
        "parts": [
          {
            "body": "$$\` \\begin{align*} g(2) &= 2 + g\\left(1\\right)
\\\\ &= 2 + 1 + g\\left(0\\right)
\\\\ &= 3 + 1
\\\\ &= 4
\\end{align*} \`

$$\` \\begin{align*} g(4) &= 2 + g\\left(2\\right)
\\\\ &= 2 + 4
\\\\ &= 6 \\; \\QED
\\end{align*} \`

$$\` \\begin{align*} g(6) &= 2 + g\\left(3\\right)
\\\\ &= 2 + 1 + g\\left(2\\right)
\\\\ &= 3 + 4
\\\\ &= 7
\\end{align*} \`

$$\` \\begin{align*} g(7) &= 1 + g\\left(6\\right)
\\\\ &= 1 + 7
\\\\ &= 8 \\; \\QED
\\end{align*} \`

$$\` \\begin{align*} g(12) &= 2 + g\\left(6\\right)
\\\\ &= 2 + 2 + g\\left(3\\right)
\\\\ &= 2 + 7
\\\\ &= 9 \\; \\QED
\\end{align*} \`",
          },
          {
            "body": "$$\` \\begin{align*} g(8) &= 2 + g\\left(4\\right)
\\\\ &= 2 + 6
\\\\ &= 8
\\end{align*} \`

Since $\`  {g(7)=g(8)=8,} \`
$\` g \`
is not a one-to-one function and $\` g \`
does not have an inverse $\` \\QED  \`",
          },
        ],
      },
    ],
  },
}
`;
